stanify/builders/stan_block_builder.py

```diff
@@ -0,0 +1,673 @@
+
+
+    def build_param_rng_functions(self, hier_est_param_names):
+
+        ignored_variables = set(self.stan_model_context.stan_data.keys()).union(
+            set(self.vensim_model_context.integ_outcome_vector_names))
+        stmt_sorter = StatementTopoSort(ignored_variables)
+
+        for sample_statements in self.stan_model_context.sample_statements:
+            stmt_sorter.add_stmt(sample_statements.lhs_variable, sample_statements.rhs_variables)
+
+        # print(stmt_sorter.dependency_graph)
+        # print(stmt_sorter.ignored_variables)
+        # print(self.stan_model_context.stan_data.keys())
+
+        param_draw_order = stmt_sorter.sort()
+        statements = self.stan_model_context.sample_statements.copy()
+
+
+        processed_statements = set()
+        for param_name in param_draw_order:
+            if param_name in self.stan_model_context.obs_integ_outcome_vector_names:
+                continue
+            for statement in statements:
+                if statement in processed_statements:
+                    continue
+                if param_name == statement.lhs_variable:
+                    if statement.init_state:
+                        param_name = param_name + "__init"
+                    if param_name in hier_est_param_names:
+
+
+                        dist_code = "rep_vector(" + f'{statement.distribution_args[0]}, R), ' + f"{', '.join(statement.distribution_args[1:])}"
+                        self.code += f"real {param_name}[R] =  {statement.distribution_type}_rng({dist_code});\n"
+                    else:
+                        self.code += f"real {param_name} = {statement.distribution_type}_rng({', '.join(statement.distribution_args)});\n"
+                    processed_statements.add(statement)
+
+    def build_obs_rng_functions(self):
+        if self.precision_context.R == 1:
+            self.code += "// Define and assign generated value to observed vector (matching vector)\n"
+
+            for statement in self.stan_model_context.sample_statements:
+                if statement.lhs_expr in self.stan_model_context.obs_integ_outcome_vector_names:
+                    vec_name = statement.lhs_expr
+                    self.code += f"array [N] real {vec_name} = {statement.distribution_type}_rng({', '.join(statement.distribution_args)});\n"
+        else:
+            self.code += "// Define observed vector (matching vector)\n"
+            for statement in self.stan_model_context.sample_statements:
+                if statement.lhs_expr in self.stan_model_context.obs_integ_outcome_vector_names:
+                    self.code += f"array[N] vector[R] {statement.lhs_expr};\n"
+
+            self.code += "// Assign generated value to observed vector (matching vector)\n"
+            self.code += "for (r in 1:R){\n"
+            self.code.indent_level += 1
+            for statement in self.stan_model_context.sample_statements:
+                #TODO @Dashadower statement.lhs_variable vs .lhs_expr
```

Comment:
> statement.lhs_variable vs .lhs_expr

The LHS of a statement is an expression, which may or may not include one or more variables. e.g,
```
my_variable + 1 ~ normal(0, 1);
```
is a valid sampling statement. Here `my_variable + 1` would be the entire expression(`lhs_expr`) and `my_variable` the variable within the expression(`lhs_variable`).

Reply:
At our level of dev, would differentiating `lhs_expr` and `lhs_variable` help or hinder? I mean couldn't we write `my_variable` ~ normal(-1,1)? There may be cases with difficult inverse functions like glm, and I wish to know some examples where `lhs_expr` is necessary.